readme.md
# HubSpot Fields JS <!-- omit in toc -->

<img src="https://img.shields.io/badge/Version-1.1.6-brightgreen" />

The package makes using JavaScript to generate HubSpot module `fields.json` files on local development straightforward.

## Table of Contents <!-- omit in toc -->

- [Requirements](#requirements)
- [Why is it needed?](#why-is-it-needed)
- [Installation](#installation)
	- [Install the package](#install-the-package)
	- [Configure Webpack Config](#configure-webpack-config)
- [Example fields.js file](#example-fieldsjs-file)
- [Usage : Field Types](#usage--field-types)
- [Usage: Groups](#usage-groups)
- [Usage: Helper Methods](#usage-helper-methods)
	- [Name + Label](#name--label)
	- [ID](#id)
	- [Default](#default)
	- [Set](#set)
	- [Visibility](#visibility)
	- [Repeater](#repeater)
	- [Help Text](#help-text)
	- [Inline Help Text](#inline-help-text)
	- [Label](#label)
	- [Required](#required)
	- [Hidden Sub Fields](#hidden-sub-fields)
	- [Locked](#locked)
	- [Choices](#choices)
	- [Children](#children)
- [Advanced Usage](#advanced-usage)
- [About iGoMoon](#about-igomoon)
	- [We're Hiring](#were-hiring)
- [Contributing](#contributing)
- [License](#license)

## Requirements

This is a webpack plugin designed to be a companion for (but not dependant on) the Hubspot Webpack Plugin created by our friends at HubSpot (Found here: [@hubspot/webpack-cms-plugins](https://www.npmjs.com/package/@hubspot/webpack-cms-plugins)).

## Why is it needed?

If you're here, then you probably love HubSpot almost as much as we do, and when creating and maintaining the `fields.json` files for modules, they can quickly become long, arduous to read and difficult to update over time. "Fields JS" gives you the ease of working with webpack with your HubSpot modules.

**How does it work?**

Within your modules, you create a `fields.js` file in place of `fields.json` and during compilation, the plugin will convert your `fields.js` files to `fields.json` that can be uploaded to your HubSpot Account. After copying to the dist folder, the plugin will search your "dist" folder for `fields.js` and then replace those with transformed `fields.json` files.

## Installation

### Install the package

```shell
npm install @igomoon/hubspot-fields-js --save-dev
```

### Configure Webpack Config

1. Set up a hubspot.config.yml using the HubSpot CMS local development instructions.
2. Add the **hubspot-fields-js** plugin to your webpack.config.js.
   - **Options**
     - `src`: The starting directory to search for field.js files eg. "./src"
     - `extraDirsToWatch`: an array of relative paths to directories that will trigger a recompilation. Useful for FieldJS partials
     - `ignore`: An array of relative paths to ignore when searching for field.js files

Example `webpack.config.js`

```javascript
const HubSpotAutoUploadPlugin = require("@hubspot/webpack-cms-plugins/HubSpotAutoUploadPlugin");
const CopyWebpackPlugin = require("copy-webpack-plugin");
// STEP 1 : Include the Fields JS Plugin
const { FieldsPlugin } = require("@igomoon/hubspot-fields-js");

module.exports = ({ account, autoupload }) => ({
  entry: "./src/index.js",
  output: {
    filename: "js/main.js",
  },
  module: {
    rules: [
		//...
	]
  },
  plugins: [

    // STEP 2 : Add to the top of  your plugins
    new FieldsPlugin({
		src: "",
		extraDirsToWatch : ["./src/fields"],
		ignore: []
	}),

    new HubSpotAutoUploadPlugin({
      autoupload,
      account,
      src: "dist",
      dest: "my-project",
    }),

    new CopyWebpackPlugin([
      { from: "src/images", to: "images" },
      { from: "src/modules", to: "modules" },
      { from: "src/templates", to: "templates" },
    ]),

  ],
});
```
1. Run webpack as normal to compile your project and convert `fields.js` to `fields.json`.


## Example fields.js file

Example `fields.js` within a module.
```javascript
const { Field } = require("@igomoon/hubspot-fields-js");

module.exports = [

	Field.text()
		.name("title","Title")
		.default("Text"),
	
	Field.richText()
		.name("text","Text")
		.default("<p>This is just some placeholder text in a rich text field</p>")

]; 
```

## Usage : Field Types

https://developers.hubspot.com/docs/cms/building-blocks/module-theme-fields

### Blog Field <!-- omit in toc -->

https://developers.hubspot.com/docs/cms/building-blocks/module-theme-fields#blog

```javascript
Field.blog()
	.name("blog_field","Blog Field")
```

### Boolean Field <!-- omit in toc -->

https://developers.hubspot.com/docs/cms/building-blocks/module-theme-fields#boolean

```javascript
Field.boolean()
	.name("boolean_field","Boolean Field")
```

Setting additional options with `set` helper method

```javascript
Field.boolean()
	.name("boolean_field","Boolean Field")
	.set("display","toggle")
	.default(false)
```

### Choice Field <!-- omit in toc -->

https://developers.hubspot.com/docs/cms/building-blocks/module-theme-fields#choice

```javascript
Field.choice()
	.name("choice_field","Choice Field")
	.choices([])
```

Setting additional options with `set` helper method

```javascript
Field.choice()
	.name("choice_field","Choice Field")
	.choices([
		["choice-1", "Choice 1"]
	])
	.set("display","select")
	.default("choice-1")
```

### Color Field <!-- omit in toc -->

https://developers.hubspot.com/docs/cms/building-blocks/module-theme-fields#color

```javascript
Field.color()
	.name("color_field","color Field")
	.default("#ffffff")
```

Setting additional options with `set` helper method

```javascript
Field.color()
	.name("color_field","color Field")
	.set("default",{
		"color": "#ffffff",
		"opacity": 80
	})
```

### CTA Field <!-- omit in toc -->

https://developers.hubspot.com/docs/cms/building-blocks/module-theme-fields#cta

```javascript
Field.cta()
	.name("cta_field","CTA Field")
```

### CRM Object Field <!-- omit in toc -->

https://developers.hubspot.com/docs/cms/building-blocks/module-theme-fields#crm-object

```javascript
Field.crmObject()
	.name("crmobject_field","CRM Object Field")
```

Setting additional options with `set` helper method

```javascript
Field.crmObject()
	.name("crmobject_field","CRM Object Field")
	.set("object_type","CONTACT")
	.set("properties_to_fetch",[])
```

### Date Field <!-- omit in toc -->

https://developers.hubspot.com/docs/cms/building-blocks/module-theme-fields#date

```javascript
Field.date()
	.name("date_field","Date Field")
```

### Date and Time Field <!-- omit in toc -->

https://developers.hubspot.com/docs/cms/building-blocks/module-theme-fields#date-and-time

```javascript
Field.dateTime()
	.name("datetime_field","Date and Time Field")
```

Setting additional options with `set` helper method

```javascript
Field.dateTime()
	.name("datetime_field","Date and Time Field")
	.set("step",30)
```

### Email Field <!-- omit in toc -->

https://developers.hubspot.com/docs/cms/building-blocks/module-theme-fields#email

```javascript
Field.email()
	.name("email_field","Email Field")
```

### Embed Field <!-- omit in toc -->

https://developers.hubspot.com/docs/cms/building-blocks/module-theme-fields#embed

```javascript
Field.embed()
	.name("embed_field","Embed Field")
```

Setting additional options with `set` helper method

```javascript
Field.embed()
	.name("embed_field","Embed Field")
	.set("supported_source_types",[ "oembed", "html" ])
	.set("supported_oembed_types",[ [ "photo", "video", "link", "rich" ] ])
```

### File Field <!-- omit in toc -->

https://developers.hubspot.com/docs/cms/building-blocks/module-theme-fields#file

```javascript
Field.file()
	.name("file_field","File Field")
```

Setting additional options with `set` helper method

```javascript
Field.file()
	.name("file_field","File Field")
	.set("picker","document")
```

### Followup Field <!-- omit in toc -->

https://developers.hubspot.com/docs/cms/building-blocks/module-theme-fields#followup-email

```javascript
Field.followUpEmail()
	.name("followupemail_field","Followup Field")
```

### Font Field <!-- omit in toc -->

https://developers.hubspot.com/docs/cms/building-blocks/module-theme-fields#font

```javascript
Field.font()
	.name("font_field","Font Field")
```

Setting additional options with `set` helper method

```javascript
Field.font()
	.name("font_field","Font Field")
	.set("load_external_fonts", true)
	.set("default",{
		"size": 12,
		"font": "Merriweather",
		"font_set": "GOOGLE",
		"size_unit": "px",
		"color": "#000",
		"styles": {}
	})
```

### Form Field <!-- omit in toc -->

https://developers.hubspot.com/docs/cms/building-blocks/module-theme-fields#form

```javascript
Field.form()
	.name("form_field","Form Field")
	.default("form_id")
```

Setting additional options with `set` helper method

```javascript
Field.form()
	.name("form_field","Form Field")
	.set("default",{
		"response_type": "inline",
		"message": "Thanks for submitting the form."
	})
```

### HubDB Row Field <!-- omit in toc -->

https://developers.hubspot.com/docs/cms/building-blocks/module-theme-fields#hubdb-row

```javascript
Field.hubdbRow()
	.name("hubdbrow_field","HubDB Row Field")
	.set("table_name_or_id","4357464")
	.default("row_id")
```

Setting additional options with `set` helper method

```javascript
Field.hubdbRow()
	.name("hubdbrow_field","HubDB Row Field")
	.set("table_name_or_id","4357464")
	.set("columns_to_fetch",[])
	.set("display_columns",[])
	.set("display_format","%0 - %1")
```

### HubDB Table Field <!-- omit in toc -->

https://developers.hubspot.com/docs/cms/building-blocks/module-theme-fields#hubdb-table

```javascript
Field.hubdbTable()
	.name("hubdbtable_field","HubDB Table Field")
```

### Icon Field <!-- omit in toc -->

https://developers.hubspot.com/docs/cms/building-blocks/module-theme-fields#icon

```javascript
Field.icon()
	.name("icon_field","Icon Field")
```

Setting additional options with `set` helper method

```javascript
Field.icon()
	.name("icon_field","Icon Field")
	.set("icon_set", "fontawesome-5.14.0")
	.set("default",{
		"name" : "accessible-icon",
		"unicode" : "f368",
		"type" : "REGULAR"
	})
```

### Image Field <!-- omit in toc -->

https://developers.hubspot.com/docs/cms/building-blocks/module-theme-fields#image

```javascript
Field.image()
	.name("image_field","Image Field")
	.default("image_field","Image Field")
```

Setting additional options with `set` helper method

```javascript
Field.icon()
	.name("image_field","Image Field")
	.set("responsive",true)
	.set("resizable",true)
	.set("show_loading",false)
	.set("default",{
		"size_type" : "auto",
		"src" : "",
		"alt" : null,
		"loading": "disabled"
	})
```

### Link Field <!-- omit in toc -->

https://developers.hubspot.com/docs/cms/building-blocks/module-theme-fields#link

```javascript
Field.link()
	.name("link_field","Link Field")
	.default("https://developers.hubspot.com/docs/cms/building-blocks/module-theme-fields")
```

Setting additional options with `set` helper method

```javascript
Field.link()
	.name("link_field","Link Field")
	.set("supported_types",["EXTERNAL", "CONTENT", "FILE", "EMAIL_ADDRESS", "BLOG"])
	.set("default",{
		"url" : {
			"content_id" : null,
			"type" : "EXTERNAL",
			"href" : ""
		},
		"open_in_new_tab" : false,
		"no_follow" : false
	})
```

### Logo Field <!-- omit in toc -->

https://developers.hubspot.com/docs/cms/building-blocks/module-theme-fields#logo

```javascript
Field.logo()
	.name("logo_field","Logo Field")
	.default("https://developers.hubspot.com/docs/cms/building-blocks/module-theme-fields")
```

Setting additional options with `set` helper method

```javascript
Field.logo()
	.name("logo_field","Logo Field")
	.set("default",{
		"override_inherited_src" : false,
		"src" : null,
		"alt" : null
	})
```

### Menu Field <!-- omit in toc -->

https://developers.hubspot.com/docs/cms/building-blocks/module-theme-fields#menu

```javascript
Field.menu()
	.name("menu_field","Menu Field")
```

### Number Field <!-- omit in toc -->

https://developers.hubspot.com/docs/cms/building-blocks/module-theme-fields#number

```javascript
Field.number()
	.name("number_field","Number Field")
```

Setting additional options with `set` helper method

```javascript
Field.number()
	.name("number_field","Number Field")
	.set("display","slider")
	.set("step",1)
	.set("min",0)
	.set("max",100)
	.set("prefix","$")
	.set("suffix","K")
```

### Page Field <!-- omit in toc -->

https://developers.hubspot.com/docs/cms/building-blocks/module-theme-fields#page

```javascript
Field.page()
	.name("page_field","Page Field")
```

Setting additional options with `set` helper method

```javascript
Field.page()
	.name("page_field","Page Field")
	.set("placeholder","Page to pull from")
```

### Rich Text Field <!-- omit in toc -->

https://developers.hubspot.com/docs/cms/building-blocks/module-theme-fields#rich-text

```javascript
Field.richText()
	.name("richtext_field","Rich Text Field")
```

Setting additional options with `set` helper method

```javascript
Field.richText()
	.name("richtext_field","Rich Text Field")
	.set("enabled_features",["bold","link","image"])
```

### Simple menu Field <!-- omit in toc -->

https://developers.hubspot.com/docs/cms/building-blocks/module-theme-fields#simple-menu

```javascript
Field.simpleMenu()
	.name("simplemenu_field","Simple menu Field")
```

### Tag Field <!-- omit in toc -->

https://developers.hubspot.com/docs/cms/building-blocks/module-theme-fields#tag

```javascript
Field.tag()
	.name("tag_field","Tag Field")
```

Setting additional options with `set` helper method

```javascript
Field.tag()
	.name("tag_field","Tag Field")
	.set("tag_value","SLUG")
```

### Text Field <!-- omit in toc -->

https://developers.hubspot.com/docs/cms/building-blocks/module-theme-fields#text

```javascript
Field.text()
	.name("text_field","Text Field")
```

Setting additional options with `set` helper method

```javascript
Field.text()
	.name("text_field","Text Field")
	.set("placeholder","Enter text here")
	.set("allow_new_line",false)
	.set("show_emoji_picker",false)
	.set("validation_regex","")
```

### URL Field <!-- omit in toc -->

https://developers.hubspot.com/docs/cms/building-blocks/module-theme-fields#url

```javascript
Field.url()
	.name("url_field","URL Field")
```

Setting additional options with `set` helper method

```javascript
Field.url()
	.name("url_field","URL Field")
	.set("supported_types",["EXTERNAL", "CONTENT", "FILE", "EMAIL_ADDRESS", "BLOG"])
	.set("default",{
		"content_id" : null,
		"href" : "http://example.com",
		"type" : "EXTERNAL"
	})
```


### Video Field <!-- omit in toc -->

https://developers.hubspot.com/docs/cms/building-blocks/module-theme-fields#video

```javascript
Field.video()
	.name("video_field","Video Field")
```

Setting additional options with `set` helper method

```javascript
Field.video()
	.name("video_field","Video Field")
	.set("default",{
		"player_id" : 32173842991,
		"height" : 1224,
		"width" : 1872,
		"conversion_asset" : {
			"type" : "CTA",
			"id" : "c3e4fa03-2c69-461d-b9af-22b2fde86bc7",
			"position" : "POST"
		}
	})
```


### Salesforce Campaign Field <!-- omit in toc -->

```javascript
Field.sfCampaign()
	.name("sfdc_campaign","Salesforce Campaign Field")
```

## Usage: Groups

https://developers.hubspot.com/docs/cms/building-blocks/module-theme-fields-overview#field-groups

You can create a field group (and nested groups) like this:

```javascript
const { Group, Field } = require("@igomoon/hubspot-fields-js");

new Group()
	.name("group", "Group")
	.children([	
		Field.text(),		
		// Other Fields/Groups
		new Group()
			.name("sub_content", "SubContent")
			.children([				
				Field.text(),				
				// Other Fields/Groups
			])		
	])
```

**IMPORTANT:** Make sure to pull in the `Group` class from the package at the top of you file.
`const { Group, Field } = require("@igomoon/hubspot-fields-js");` 

Looking for [Repeating Groups](#repeater)? See the [Helper Methods](#usage-helper-methods) below.
https://developers.hubspot.com/docs/cms/building-blocks/module-theme-fields-overview#repeating-groups


## Usage: Helper Methods

### Name + Label

`name(name, label)`

```javascript
Field.text()
	.name("field_name","Field Label")
```

### ID

`id(id)`

```javascript
Field.text()
	.id("321732-3231-3213554-21")
```

### Default

`default(value)`

```javascript
Field.text()
	.default("value to set")
```

### Set

`set(key,value)`

```javascript
Field.text()
	.set("placeholder","Placeholder Text")
```

### Visibility

`visibleIf(controllingField, regex, type = "EQUAL")`

Sets controlling field options under visibility

```javascript
Field.text()
	.visibleIf("other_field_id","value to test")
```

```javascript
Field.text()
	.visibleIf("other_field_id","value to test","NOT_EQUAL")
```

### Repeater

https://developers.hubspot.com/docs/cms/building-blocks/module-theme-fields-overview#repeaters

`repeat(overrides = {})`

```javascript
Field.text()
	.repeat()
```

```javascript
new Group()
	.name("group","Group")
	.children([])
	.repeat()
```

Override the repeat options like so:

```javascript
new Group()
	.name("group","Group")
	.children([])
	.repeat({
		"min": null,
		"max": 5,
		"sorting_label_field": 'group.name',
		"default": 1
	})
```

### Help Text

`helpText(text)`

```javascript
Field.text()
	.helpText("Help text")
```

### Inline Help Text

`inlineHelpText(text)`

```javascript
Field.text()
	.inlineHelpText("Inline Help Text")
```

### Label

`label(value)`

```javascript
Field.text()
	.label("Label Value")
```

### Required

`required(required = true)`

```javascript
Field.text()
	.required()
```

```javascript
Field.text()
	.required(false)
```

### Hidden Sub Fields

`hiddenSubfields(subFields)`

Sets hidden_subfields under visibility

```javascript
Field.text()
	.hiddenSubfields({
      "font": true,
      "size": true
    })
```

### Locked

`locked(locked = true)`

```javascript
Field.text()
	.locked()
```

```javascript
Field.text()
	.locked(false)
```

### Choices

For choice Field only
`choices(choices = [])`

```javascript
Field.choice()
	.choices([])
```

### Children

For Groups Only
`children(children = [])`

```javascript
new Group()
	.children([])
```

## Advanced Usage

What's the best feature of all? Hubspot Fields JS allows you to abstract your fields into partials. Giving you the ability to create common fields once and then pull them into any Fields-JS enabled module that you like. Removing the pain of constant duplication and allowing you to make global changes by just changing a sinle partial file.

The example below creates a text field partial with a validation regex for an id attribut, then in 2 different module fields.js file, that partial is pulled in and inserted. 

`src/fields/cssID.js`

```javascript
const { Field } = require('@igomoon/hubspot-fields-js');
module.exports = function () {
	return Field
		.text()
		.name('css_id', 'CSS ID');
		.set("validation_regex","-?[a-zA-Z]+[a-zA-Z0-9-]+")
}
```

`src/modules/example.module/fields.js`

```javascript
const { Field } = require("@igomoon/hubspot-fields-js");

// Pull in the Partial (IMPROTANT: Relative to the modules final destination in the dist folder) 
let cssID = require('../../../src/fields/cssID');

module.exports = [
	Field.text()
		.name("title","Title")
		.default("Text"),
	// This will insert the CSS ID field for this module
	cssID()
]; 
```


`src/modules/second-example.module/fields.js`

```javascript
const { Group, Field } = require("@igomoon/hubspot-fields-js");

// Pull in the Partial (IMPROTANT: Relative to the modules final destination in the dist folder) 
let cssID = require('../../../src/fields/cssID');

module.exports = [
	new Group()
		.name("group","Group")
		.children([
			Field.text()
				.name("title","Title")
				.default("Text"),			
			Field.richText()
				.name("text","Text")
				.default("<p>This is just some placeholder text in a rich text field</p>")
		])
	// This will insert the CSS ID field for this module
	cssID()
]; 
```

## About iGoMoon

[iGoMoon](https://www.igomoon.com/)’s mission is to help companies grow through data-driven communication, smart technology and relevant content. We live and breathe HubSpot and being Certified in Advanced CMS Implementation we are constantly improving our development process to better serve our customers. 

Get to know us more on our [Vlog Månresan (The moonshot journey)](https://www.igomoon.com/manresan)

### We're Hiring

We are always looking for skilled developers at iGoMoon, if you have HubSpot experience then there may be a spot for you here. Think you have what it takes? Then try our [HubSpot Developer Coding Task](https://github.com/iGoMoon/developer-coding-task) and connect with us on our career page here: [https://careers.igomoon.com/connect](https://careers.igomoon.com/connect)

## Contributing

For more information on developing, see the [Contributing Guide](CONTRIBUTING.md).

## License

Apache 2.0


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [1.2.0] - 2022-01-01
- Js Partials are now imported relative to src file, NOT from final dist file. Field JS tranformation using the src file and then updates webpack's emission functionlity to properly handle the upload.
- Plugin now accepts options
     - `src`: The starting directory to search for field.js files eg. "./src"
     - `extraDirsToWatch`: an array of relative paths to directories that will trigger a recompilation. Useful for FieldJS partials
     - `ignore`: An array of relative paths to ignore when searching for field.js files  

	```javascript
	new FieldsPlugin({
		"src": "",
		"extraDirsToWatch" : ["./src/fields"],
		"ignore": []
	})
	```

## [1.1.6] - 2021-11-11
- Allow null values in array and filter them out of the final JSON
- `applyIf` helper method
- `parseArgs` helper method

## [1.1.5] - 2021-05-11
- Hot fixes

## [1.1.4] - 2021-05-10
- Need some Defaults
 
## [1.1.3] - 2021-05-07
- Choices and Children helpers
 
## [1.1.2] - 2021-05-07
- Wrong version in ReadMe.md

## [1.1.1] - 2021-05-07 (Not published on NPM)
- Upadate this changelog

## [1.1.0] - 2021-05-07 (Not published on NPM)
- Update Readme and contributing file

## [1.0.4] - 2021-05-07
- Update Field methods so we dont override hubspot defaults

## [1.0.3] - 2021-02-25
- New Field Helper Functions: crmObject, hubdbRow, embed, video, sfCampaign

## [1.0.2] - 2021-02-16
- Package.json updates

## [1.0.1] - 2021-02-16
- Remove unnecessary dependencies

## [1.0.0] - 2021-02-16
- Initial Publish


src/fields/Field.js
const ModuleField = require('./ModuleField');

class Field extends ModuleField {

    constructor(data) {
        super();
        this.data = Object.assign({
			"required": false,
			"locked": false,
			"help_text": "",
			"inline_help_text": "",
			//"default": null (HS handles this for us)
		}, data);
    }

    /**
     * TODO Add Styles Fields for
     * Style Groups 
     */

    /**
     * Blog field
     * @param {Object} overrides 
     */
    static blog(overrides) {
        return new Field(Object.assign({
            "name": "blog_field",
            "label": "Blog field",
            "type": "blog"
        }, overrides));
    }

    /**
     * Boolean field
     * @param {Object} overrides 
     */
    static boolean(overrides) {
        return new Field(Object.assign({
            "name": "boolean_field",
            "label": "Boolean field",
            "type": "boolean",
            //"default": false
        }, overrides));
    }

    /**
     * Choice field
     * @param {Object} overrides 
     * @param {Array} choices 
     */
    static choice(overrides, choices = []) {
        return new Field(Object.assign({
            "name": "choice_field",
            "label": "Choice field",
            //"display": "select",
            "choices": choices,
            "type": "choice"
        }, overrides));
    }

    /**
     * Color field
     * @param {Object} overrides 
     */
    static color(overrides) {
        return new Field(Object.assign({
            "name": "color_field",
            "label": "Color field",
            "type": "color"
        }, overrides));
    }

    /**
     * Cta field
     * @param {Object} overrides 
     */
    static cta(overrides) {
        return new Field(Object.assign({
            "name": "cta_field",
            "label": "CTA field",
            "type": "cta"
        }, overrides));
	} 
	
	/**
     * CRM Object field
     * @param {Object} overrides
	 * @param {String} overrides.object_type - Type of CRM Object the user can pick from.
  	 * @param {Array} overrides.properties_to_fetch - Array of property names associated with the object type in string form. {@link https://developers.hubspot.com/docs/cms/features/custom-objects#supported-crm-object-types}
  	 * @param {Object} overrides.default - Object with id of default selected object instance. Contact ID, Company ID etc
	 *
	 * @see {@link https://developers.hubspot.com/docs/cms/building-blocks/module-theme-fields#crm-object} 
     */
    static crmObject(overrides) {
        return new Field(Object.assign({
            "name" : "crmobject_field",
			"label" : "CRM object",
			"object_type" : "CONTACT",
			//"properties_to_fetch" : [ ],
			"type" : "crmobject"
        }, overrides));
    }

    /**
     * Date field
     * @param {Object} overrides 
     */
    static date(overrides) {
        return new Field(Object.assign({
            "name": "date_field",
            "label": "Date field",
            "type": "date"
        }, overrides));
    }

    /**
     * DateTime field
     * @param {Object} overrides 
     */
    static dateTime(overrides) {
        return new Field(Object.assign({
            "name": "datetime_field",
			"label": "Date and time field",
			//"step" : 30,
            "type": "datetime"
        }, overrides));
    }

    /**
     * Email field
     * @param {Object} overrides 
     */
    static email(overrides) {
        return new Field(Object.assign({
            "name": "email_field",
            "label": "Email Address Field",
            "type": "email"
        }, overrides));
    }
	
	/**
     * Embed field
     * @param {Object} overrides
	 * @param {Array} overrides.supported_source_types - Supported source types for either oEmbed URLs or HTML embed code
  	 * @param {Array} overrides.supported_oembed_types - Supported oEmbed type including "photo", "video", "link", and "rich".
  	 * @param {Object} overrides.default - An array containing the "source_type" parameter. This parameter has one string based value from the options provided in the "supported_source_types" parameter.
	 *
	 * @see {@link https://developers.hubspot.com/docs/cms/building-blocks/module-theme-fields#embed} 
     */
    static embed(overrides) {
        return new Field(Object.assign({
            "name" : "embed_field",
			"label" : "Embed",
			//"supported_source_types" : [ "oembed", "html" ],
			//"supported_oembed_types" : [ "photo", "video", "link", "rich" ],
			"type" : "embed",
        }, overrides));
    }

    /**
     * File field
     * @param {Object} overrides 
     */
    static file(overrides) {
        return new Field(Object.assign({
            "name": "file_field",
			"label": "File field",
			//"picker" : "file",
            "type": "file"
        }, overrides));
    }

    /**
     * Follow up email field
     * @param {Object} overrides 
     */
    static followUpEmail(overrides) {
        return new Field(Object.assign({
            "name": "followupemail_field",
            "label": "Followup email field",
            "type": "followupemail"
        }, overrides));
    }

    /**
     * Font field
     * @param {Object} overrides 
     */
    static font(overrides) {
        return new Field(Object.assign({
            "name": "font_field",
            "label": "font_field",
			"type": "font",
			//"load_external_fonts" : true,
			//"default" : {
			// 	"size": 12,
			// 	"size_unit": "px",
			// 	"color": "#000",
			// 	"styles": {}
			// }
        }, overrides));
    }

    /**
     * Form field
     * @param {Object} overrides 
     */
    static form(overrides) {
        return new Field(Object.assign({
            "name": "form_field",
            "label": "Form field",
            "type": "form",
            // "default": {
            //     "response_type": "inline",
            //     "message": "Thanks for submitting the form."
            // }
        }, overrides));
	}
		
	/**
     * HubDB Row field
     * @param {Object} overrides
	 * @param {string} overrides.table_name_or_id - The name or ID of the HubDB table. This field is required.
  	 * @param {Array} overrides.columns_to_fetch - An array of column names to fetch from the table. If left blank, will return all columns in the table.
  	 * @param {Array} overrides.display_columns - An array of column names to use in choice label. If left blank, will return only the first column in the table.
  	 * @param {String} overrides.display_format - The format you would like the column data to display in the HubDB row selector using the percent symbol and number to designate a column. 
  	 * @param {Object} overrides.default - Object containing “id” for setting the default hubdb row.
	 *
	 * @see {@link https://developers.hubspot.com/docs/cms/building-blocks/module-theme-fields#hubdb-row} 
     */
	static hubdbRow(overrides) {
        return new Field(Object.assign({
            "name" : "hubdbrow_field",
			"label" : "HubDB row",
			"table_name_or_id" : "",
			// "columns_to_fetch" : [],
			// "display_columns" : [],
			// "display_format" : "%0 - %1",
			"type" : "hubdbrow",
        }, overrides));
    }

    /**
     * HubDB Table field
     * @param {Object} overrides 
     */
    static hubdbTable(overrides) {
        return new Field(Object.assign({
            "name": "hubdbtable_field",
            "label": "HubDB table field",
            "type": "hubdbtable"
        }, overrides));
    }

    /**
     * Icon field
     * @param {Object} overrides 
     */
    static icon(overrides) {
        return new Field(Object.assign({
            "name": "icon_field",
            "label": "Icon field",
			"type": "icon",
			"default": {}
        }, overrides));
    }

    /**
     * Image field
     * @param {Object} overrides 
     */
    static image(overrides) {
        return new Field(Object.assign({
            "name": "image_field",
            "label": "Image field",
            //"responsive": true,
            //"show_loading": false,
            //"resizable": true,
			"type": "image",
			"default" : {
				"size_type" : "auto",
				"src" : "",
				"alt" : null,
				"loading": "disabled"
			}
        }, overrides));
    }

    /**
     * Link field
     * @param {Object} overrides 
     */
    static link(overrides) {
        return new Field(Object.assign({
            "name": "link_field",
            "label": "Link field",
            "type": "link",
            //"supported_types": ["EXTERNAL", "CONTENT", "FILE", "EMAIL_ADDRESS", "BLOG"],
            "default": {
				"url" : {
					"content_id" : null,
					"type" : "EXTERNAL",
					"href" : ""
				},
				"open_in_new_tab" : false,
				"no_follow" : false
			}
        }, overrides));
    }

    /**
     * Logo field
     * @param {Object} overrides 
     */
    static logo(overrides) {
        return new Field(Object.assign({
            "name": "logo_field",
            "label": "Logo field",
            "type": "logo",
            "default": {
				"override_inherited_src" : false,
				"src" : null,
				"alt" : null
			}
        }, overrides));
    }

    /**
     * Menu field
     * @param {Object} overrides 
     */
    static menu(overrides) {
        return new Field(Object.assign({
            "name": "menu_field",
            "label": "Menu field",
            "type": "menu",
            "default": "default"
        }, overrides));
    }

    /**
     * Text field
     * @param {Object} overrides 
     */
    static number(overrides) {
        return new Field(Object.assign({
            "name": "number_field",
            "label": "Number field",
            "type": "number",
            // "display": "text",
            // "step": 1,
            // "min": null,
            // "max": null
        }, overrides));
    }

    /**
     * Page field
     * @param {Object} overrides 
     */
    static page(overrides) {
        return new Field(Object.assign({
            "name": "page_field",
            "label": "Page field",
            "type": "page"
        }, overrides));
    }

    /**
     * Rich text field
     * @param {Object} overrides 
     */
    static richText(overrides) {
        return new Field(Object.assign({
            "name": "richtext_field",
            "label": "Rich text field",
            "type": "richtext"
        }, overrides));
    }

    /**
     * Simple field
     * @param {Object} overrides 
     */
    static simpleMenu(overrides) {
        return new Field(Object.assign({
            "name": "simplemenu_field",
            "label": "Simple menu field",
            "type": "simplemenu"
        }, overrides));
    }

    /**
     * Tag field
     * @param {Object} overrides 
     */
    static tag(overrides) {
        return new Field(Object.assign({
            "name": "tag_field",
            "label": "Tag field",
            //"tag_value": "SLUG",
            "type": "tag"
        }, overrides));
    }

    /**
     * Text field
     * @param {Object} overrides 
     */
    static text(overrides) {
        return new Field(Object.assign({
            "name": "text",
            "label": "Text",
            // "validation_regex": "",
            // "allow_new_line": false,
            // "show_emoji_picker": false,
            "type": "text"
        }, overrides));
    }

    /**
     * URL field
     * @param {Object} overrides 
     */
    static url(overrides) {
        return new Field(Object.assign({
            "name": "url_field",
            "label": "URL field",
            "type": "url",
            //"supported_types": ["EXTERNAL", "CONTENT", "FILE", "EMAIL_ADDRESS", "BLOG"],
            "default": {
				"content_id" : null,
				"href" : "",
				"type" : "EXTERNAL"
			}
        }, overrides));
    }
	
	/**
     * Video field
     * @param {Object} overrides
  	 * @param {Object} overrides.default - Video object with settings for player_id, height, width, size_type, and conversion_asset object.
	 *
	 * @see {@link https://developers.hubspot.com/docs/cms/building-blocks/module-theme-fields#video} 
     */
    static video(overrides) {
		return new Field(Object.assign({
			"name" : "videoplayer_field",
			"label" : "Video",
			"type" : "videoplayer"
        }, overrides));
	}
	
	/**
     * Salesforce Campaign field
     * @param {Object} overrides
     */
    static sfCampaign(overrides) {
		return new Field(Object.assign({
			"name" : "sfdc_campaign",
			"label" : "Salesforce Campaign Field",
			"type" : "salesforcecampaign",
        }, overrides));
	}
	
    /**
     * Return the fields Json representation.
     */
    toJSON() {
        return this.data;
    }

    /**
     * Set default value of field
     * @param {*} value 
     */
    default(value) {
        switch (this.data.type) {
            case "link":
			case "url":
                this.data.default.url.href = value;
                break;
            case "color":
                this.data.default.color = value;
                break;
            case "image":
                this.data.default.src = value;
                break;
            case "form":
                this.data.default.form_id = value;
                break;
            case "crmobject":
            case "hubdbrow":
				this.data.default.id = value;
                break;
            case "embed":
				this.data.default.source_type = value;
                break;
            default:
                this.data.default = value;
                break;
        }

        return this;
    }

}

module.exports = Field;